marketplace/client.py

```
@@ -6,16 +6,61 @@
 """
 
 import os
+from functools import wraps
 from urllib.parse import urljoin
 
 import requests
+from keycloak import KeycloakOpenID
 from requests import Response
 
 from .version import __version__
 
 MP_DEFAULT_HOST = "https://materials-marketplace.eu/"
 
 
+def configure_token(func):
+    @wraps(func)
+    def func_(self, *arg, **kwargs):
+        r = func(self, *arg, **kwargs)
+        if r.status_code == 401:
+            token = configure()
+            if token is None:
+                raise (
+                    "Authentication failure. Please provide valid MP_ACCESS_TOKEN or keycloak credentials."
+                )
+            os.environ["MP_ACCESS_TOKEN"] = token
+            self.access_token = token
+
+        r = func(self, *arg, **kwargs)
+        if r.status_code > 400:
+            raise Exception("Server returned with an Exception. Details: ", r.text)
+        return r
+
+    return func_
+
+
+def configure():
+    # Configure client
+    server_url = os.environ.get("KEYCLOAK_SERVER_URL")
+    client_id = os.environ.get("KEYCLOAK_CLIENT_ID")
+    realm_name = os.environ.get("KEYCLOAK_REALM_NAME")
+    client_key = os.environ.get("KEYCLOAK_CLIENT_SECRET_KEY")
+    user = os.environ.get("MARKETPLACE_USERNAME")
+    passwd = os.environ.get("MARKETPLACE_PASSWORD")
+    keycloak_openid = KeycloakOpenID(
+        server_url=server_url,
+        client_id=client_id,
+        realm_name=realm_name,
+        client_secret_key=client_key,
+    )
+    try:
+        token = keycloak_openid.token(user, passwd)
+        token = token["access_token"]
+        return token
+    except Exception:
+        return None
```

This means _all_ errors will return `None`, which in your `configure_token` will trigger an authentication error, right?
- Are you sure those are all the possible errors? What is keycloak is down, for example?
- I would handle exceptions properly and propragate when necessary, rather than catching -> convert to None -> raise new exception

---

Now I catch all exceptions and display a generic message "Authentication failure" along with actual reason for the error. Any error other than authentication will be caught separately and displayed to user.